File: ruby/src/best-time-to-buy-and-sell-stock-with-cooldown.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  recur(prices, Array.new(prices.size) { Array.new(2) }, 0, 1)
end

def recur(prices, cache, i, can_buy)
  return 0 if i >= prices.size
  cache[i][can_buy] ||= begin
    if can_buy == 1
      bought = recur(prices, cache, i + 1, 0) - prices[i]
      cooled = recur(prices, cache, i + 1, 1)
      [bought, cooled].max
    else
      sold = recur(prices, cache, i + 2, 1) + prices[i]
      cooled = recur(prices, cache, i + 1, 0)
      [sold, cooled].max
    end
  end
end

File: elixir/best-time-to-buy-and-sell-stock-with-cooldown.ex
defmodule Solution do
  @spec max_profit(prices :: [integer]) :: integer
  def max_profit(prices), do: recur(%{}, prices, 1)

  @spec recur(cache :: %{[integer] => %{ integer => integer }}, prices :: [integer], can_buy :: integer) :: integer
  def recur(cache, [], _), do: 0
  def recur(cache, prices, can_buy) when cache[prices][can_buy], do: cache[prices][can_buy]
  def recur(cache, prices, 1), do: Map.put(cache, prices, max(recur(cache, tl(prices), 0) - hd(prices), recur(cache, tl(prices), 1)))
  def recur(cache, prices, 0), do: Map.put(cache, prices, max(recur(cache, tl(tl(prices)), 1) + hd(prices), recur(cache, tl(prices), 0)))
end
